### DISCUSSION TITLE
llvm-mos life example running so sloooow

### DISCUSSION THREAD

**Discussion by WojciechGw** [picocomputer/community#85] [Q&A]
llvm-mos life example from newest video works 
but running very slow
any ideas ?

**Comment by rumbledethumps** [comment]
It's slow. Best results will come from LLVM-MOS with Release optimizations. cc65 is unbearable.

**Reply by WojciechGw** [reply]
so I did

ooooo you sped up the video at the end ... "editing magic" :) now I get it :)))

**Reply by rumbledethumps** [reply]
There's opportunity to speed up the algorithm. Take advantage of the fact that most of the screen is dead cells. You can also play with other starting generations.

**Reply by netzerohero** [reply]
Great video; helpful in gaining understanding of the video hardware and cross boundary memory read/write access methods (xram / 6502-ram). Enjoy thinking about in the context of the classic game-of-life.  As implemented, display-pixels are 1bpp; multiple byte-accesses are needed for read-modify-write at the bit-level.  I'm wondering if there would be an appreciable speed advantage working only at the byte-level, 8bpp, especially in the set & unset functions: both for the RIA xram-address calculation: RIA.addr0 = (x / 8) + (320 / 8 * y); and for the read-modify-writes: RIA.rw0 |= bit  -or-  RIA.rw0 &= ~bit.

**Comment by grakoczy** [comment]
Hi,
I've modified the code from this great video: [https://www.youtube.com/watch?v=ndAfWKmKF34&t=713s](https://www.youtube.com/watch?v=ndAfWKmKF34&t=713s) for RP6502.
I've used [vruumllc](https://github.com/vruumllc) Bitmap Graphics library to simplify things in the future.
This implementation is reasonably fast with 140x140 map size.
Here is the code for CC65: [https://github.com/grakoczy/picocomputer-game-of-life-cc65](https://github.com/grakoczy/picocomputer-game-of-life-cc65)
Video is not speed up. 

https://github.com/user-attachments/assets/b7500b9c-fcc4-4195-b751-ca09426321fe

**Comment by grakoczy** [comment]
Hi,

Finally I've been able to setup llvm-mos environment and built the code with new compiler.
It works few times faster.
New repo is here: [https://github.com/grakoczy/picocomputer-game-of-life](url)


https://github.com/user-attachments/assets/543e834b-3953-421f-8951-b1786a421e9e
